drivers/bme680/bme680_saul.c

```diff
@@ -0,0 +1,210 @@
+    _temp_valid[dev] = true;
+    _press_valid[dev] = true;
+    _hum_valid[dev] = true;
+    _gas_valid[dev] = true;
+    return BME680_OK;
+}
+
+static int read_temp(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_t *)dev);
+    if (dev_index == BME680_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_temp_valid[dev_index] || read(dev_index) == BME680_OK) {
+        /* mark local variable as invalid */
+        _temp_valid[dev_index] = false;
+
+        data->val[0] = _temp[dev_index];
+        data->unit = UNIT_TEMP_C;
+        data->scale = -2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_press(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_t *)dev);
+    if (dev_index == BME680_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_press_valid[dev_index] || read(dev_index) == BME680_OK) {
+        /* mark local variable as invalid */
+        _press_valid[dev_index] = false;
+
+        data->val[0] = _press[dev_index];
+        data->unit = UNIT_PA;
+        data->scale = 2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_hum(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_t *)dev);
+    if (dev_index == BME680_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_hum_valid[dev_index] || read(dev_index) == BME680_OK) {
+        /* mark local variable as invalid */
+        _hum_valid[dev_index] = false;
+
+        data->val[0] = _hum[dev_index];
+        data->unit = UNIT_PERCENT;
+        data->scale = -2;
+        return 1;
+    }
+    return -ECANCELED;
+}
+
+static int read_gas(const void *dev, phydat_t *data)
+{
+    /* find the device index */
+    unsigned dev_index = _dev2index((const bme680_t *)dev);
+    if (dev_index == BME680_NUMOF) {
+        /* return with error if device index could not be found */
+        return -ECANCELED;
+    }
+
+    /* either local variable is valid or fetching it was successful */
+    if (_gas_valid[dev_index] || read(dev_index) == BME680_OK) {
+        /* mark local variable as invalid */
+        _gas_valid[dev_index] = false;
+
+        if (_gas[dev_index] > INT16_MAX) {
+            data->val[0] = _gas[dev_index] / 1000;
+            data->scale = 3;
+        }
+        else {
+            data->val[0] = _gas[dev_index];
+            data->scale = 0;
+        }
+        data->unit = UNIT_OHM;
```

Comment:
Ohm?

Reply:
Yep, the gas measurement is returned as the resistance of the gas sensitive layer. What ever it means. Unfortunately, the datasheet gives no explanation or hints how the value has to be interpreted : worried:

Reply:
They just recommend to use the [BSEC](https://www.bosch-sensortec.com/software-tools/software/bsec) to get IAQ.

Reply:
I have the import of the BSEC as package on my ToDo list.

Reply:
@BoschSensortec: Care to help out here?

Reply:
The biggest chalange will be that each user of the BSCE has to accept the license before it is downloadable.